test/e2e/e2e.go

```diff
@@ -0,0 +1,164 @@
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"testing"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	metallbv1alpha "github.com/metallb/metallb-operator/api/v1alpha1"
+	testclient "github.com/metallb/metallb-operator/test/e2e/client"
+	corev1 "k8s.io/api/core/v1"
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	goclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest/printer"
+)
+
+const (
+	// MetallbNameSpace contains the name of the metal LB operator  namespace
+	MetallbNameSpace = "metallb-system"
+	// MetallbOperatorDeploymentName contains the name of the metal LB operator deployment
+	MetallbOperatorDeploymentName = "metallboperator-controller-manager"
+	// MetallbOperatorDeploymentLabel contains the label of the metal LB operator deployment
+	MetallbOperatorDeploymentLabel = "controller-manager"
+	// MetallbOperatorCRDName contains the name of the metal LB operator CRD
+	MetallbOperatorCRDName = "metallbs.metallb.io"
+	// MetallbOperatorFile contains the
+	MetallbOperatorFile = "metallb.yaml"
+)
+
+var (
+	operatorDeployed bool
+)
+
+func init() {
+	operatorDeployed = true
+}
+
+func RunE2ETests(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	RunSpecsWithDefaultAndCustomReporters(t,
+		"Controller Suite",
+		[]Reporter{printer.NewlineReporter{}})
+}
+
+var _ = Describe("validation", func() {
+	Context("general", func() {
+		It("should have the metal LB operator namespace", func() {
+			_, err := testclient.Client.Namespaces().Get(context.Background(), MetallbNameSpace, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("should have all the nodes in ready state", func() {
+			nodes, err := testclient.Client.Nodes().List(context.Background(), metav1.ListOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			for _, node := range nodes.Items {
+				nodeReady := false
+				for _, condition := range node.Status.Conditions {
+					if condition.Type == corev1.NodeReady &&
+						condition.Status == corev1.ConditionTrue {
+						nodeReady = true
+					}
+				}
+				Expect(nodeReady).To(BeTrue(), "Node ", node.Name, " is not ready")
+			}
+		})
+	})
+
+	Context("metallb", func() {
+		var expectedReadyReplicas int32
+
+		It("should have the metal LB operator deployment in running state", func() {
+			deploy, err := testclient.Client.Deployments(MetallbNameSpace).Get(context.Background(), MetallbOperatorDeploymentName, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(deploy.Status.Replicas).To(Equal(deploy.Status.ReadyReplicas))
+
+			pods, err := testclient.Client.Pods(MetallbNameSpace).List(context.Background(), metav1.ListOptions{
```

Comment:
isn't it redundant to check if the pods are running after we checked that all of the deployment's replicas are ready?

Reply:
@fedepaol What do you think?
The same thing is done in cnf-tests. For example:
https://github.com/openshift-kni/cnf-features-deploy/blob/ccfe348e2a33f8318e2f59f7873e55eb3cb51481/cnf-tests/testsuites/validationsuite/cluster/validation.go#L108

Reply:
It won't hurt I guess